### ISSUE TITLE
Order status is saved as processing instead of wc-processing in database.

### ISSUE THREAD

**Issue by lakshmana** [Twispay/WooCommerce#17]
The order status for a subscription product is saved in database as 'processing' instead of 'wc-processing'. Because of this the order confirmation email is also not sent to the customer. 
![image](https://user-images.githubusercontent.com/3132026/128507334-5fc4c5f9-0734-4bb0-95f0-72334fc3c998.png)

**Comment by lakshmana** [comment]
The code at line number 254 @ https://github.com/Twispay/WooCommerce/blob/master/Twispay%20WooCommerce%20Plugin/twispay/helpers/Twispay_TW_Status_Updater.php#L254 `$order->update_status('processing', esc_html__( $tw_lang['wa_order_status_notice'], 'woocommerce' ));` is not saving order status with wc-processing. 

Can you please check on this?

**Comment by lakshmana** [comment]
After digging hours, I found when the IPN callback method `Twispay::server_to_server` is called the `get_post_stati()` does not load the woocommerce post_status. for eg, the following are loaded during `Twispay::server_to_server` callback
```
array (size=13)
  'publish' => string 'publish' (length=7)
  'future' => string 'future' (length=6)
  'draft' => string 'draft' (length=5)
  'pending' => string 'pending' (length=7)
  'private' => string 'private' (length=7)
  'trash' => string 'trash' (length=5)
  'auto-draft' => string 'auto-draft' (length=10)
  'inherit' => string 'inherit' (length=7)
  'request-pending' => string 'request-pending' (length=15)
  'request-confirmed' => string 'request-confirmed' (length=17)
  'request-failed' => string 'request-failed' (length=14)
  'request-completed' => string 'request-completed' (length=17)
  'acf-disabled' => string 'acf-disabled' (length=12)
```
The `get_post_status` method in `Abstract_WC_Order_Data_Store_CPT` at `woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php` has the following code 

```
$valid_statuses = get_post_stati();
if ( ! in_array( $post_status, array( 'auto-draft', 'draft', 'trash' ), true ) && in_array( 'wc-' . $post_status, $valid_statuses, true ) ) {
			$post_status = 'wc-' . $post_status;
		}
```
Now since `$valid_statuses` does not have woocomerce post_status like `wc-pending, wc-processing`, the `$post_status` is not prefixed with `wc-`.

The problem seems to lie with the constructor of `Twispay` ie `add_action('woocommerce_after_register_post_type', array( $this, 'server_to_server' ) );`

**Comment by lakshmana** [comment]
~~The easiest solution seems to call `WC_Post_Types::register_post_status();` before `Twispay_TW_Status_Updater::updateStatus_IPN($orderId, $status, $tw_lang);` is called in `twispay/twispay.php`~~

The current Twispay's constructor has setup `Twispay's::server_to_server` method as a callback for `woocommerce_after_register_post_type` action. Because of this all the Woocommerce Orders and WooCommerce Subscription methods are not available at this state. 

Is there any reason why this method is hooked at woocommerce_after_register_post_type instead of 'init' for eg
`add_action('init', array( $this, 'server_to_server' ) );` ? This seems to fix on my side.

**Comment by lakshmana** [comment]
The `twispay/helpers/Twispay_TW_Status_Updater.php` helper class does not check if the order status is already complete or processing. Because of this any change in order status from WC hook `woocommerce_order_status_processing` is reverted back to processing. for eg. the following hook will change the order status to processing but the `Twispay_TW_Status_Updater` helper will revert it back to processing. 
```
add_action( 'woocommerce_order_status_processing', function($order_id){
    if ( ! $order_id ) {
        return;
    }
    $order = wc_get_order( $order_id );
    $order->update_status( 'completed', 'auto-switched-by-script', true );
},100, 1);
```

I had to add this condition on `Twispay_TW_Status_Updater` to solve this issue. 
```
/* Mark order as completed. */
if($order->get_status() !== 'completed') {
    $order->update_status('processing', esc_html__($tw_lang['wa_order_status_notice'], 'woocommerce'));
}
```

**Comment by carco** [comment]
I cannot reproduce (also, with that condition (order !== completed) the hook will not be executed/triggered).

**Comment by carco** [comment]
(the main issue) Fixed on #18
